Common/Util/ConcurrentSet.cs

```diff
@@ -0,0 +1,227 @@
+﻿/*
+ * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
+ * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+using System;
+using System.Collections;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace QuantConnect.Util
+{
+    /// <summary>
+    /// Provides a thread-safe set collection that mimics the behavior of <see cref="HashSet{T}"/>
+    /// </summary>
+    /// <typeparam name="T">The item type</typeparam>
+    public class ConcurrentSet<T> : ISet<T>
+    {
+        private readonly ConcurrentDictionary<T, T> _set = new ConcurrentDictionary<T, T>();
```

Comment:
Think it might be cheaper memory-wise to just make the value a `byte`?

Reply:
I'm not sure that it's simple as there's struct alignment issues to consider, for example, when instantiating the `KeyValuePair<T, byte>` instances for enumeration. Also, if using `byte` we'll be subject to boxing/unboxing semantics since it's a value type (also when creating `KeyValuePair<T, byte>` instances during enumeration). The internal `ConcurrentDictionary<TKey, TValue>.Node` type _is_ a class, so IIRC, it isn't subject to the same alignment concerns as KVP, but we often enumerate this type which instantiates KVP instances, so it's worth noting.

Further, the potential difference in memory required, assuming the `byte` actually gets stored as a single byte vs a 4-byte dword (8-byte on when x64, obj reference), would be a maximum of 3 (7) bytes per entry, so even at 1M entries, we're still under 3MB (7MB) of memory saved.  Either way, I think 3 (or 7) bytes per entry is infinitesimally smaller than the actual object being stored making the savings practically negligible. Now, if it _does_ store it internally as a `byte`, then we're guaranteed to get the performance hit of boxing/unboxing the byte -- which I think is the bigger thing to balance vs a few KBs (in reality algos don't have more than a few hundred, maybe a thousand consolidators).

I would prefer to leave this as-is without evidence supporting it as an issue.

Reply:
Yeah, good points there will never be that amount of consolidators where this matters, but it might add an overhead 👍